### PR TITLE
NestedHelper shouldn't use deprecated TermInSetQuery method

### PR DESCRIPTION
**Pull request by cbuescher**
The getTermData method in TermInSetQuery is deprecated and not needed for what we do in NestedHelper. We can remove its use by using other provided methods.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/search/NestedHelper.java [hunk 1]
```diff
@@ -53,14 +53,16 @@ public boolean mightMatchNestedDocs(Query query) {
...
+        } else if (query instanceof TermInSetQuery tis) {
+            try {
+                if (tis.getTermsCount() > 0) {
+                    return mightMatchNestedDocs(tis.getField());
+                } else {
+                    return false;
+                }
+            } catch (IOException e) {
+                // this handling isn't needed any more once we move to Lucene 10
+                throw new AssertionError("We are not doing IO here, this should never happen.");
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/search/NestedHelper.java:65]: AssertionError will crash the node right? I do not like throwing this in a production code path. 

We can add an `assert false` here instead as that will only trip when assertions are enabled. 

As for what to do if we do get an IOException from `getTermsCount()` maybe rethrow as an `UncheckedIOException`

**Review by benwtrent — COMMENTED** [review]
I haven an issue with the exception handling, but the code looks good. What we did before was way too complicated :)

File: server/src/main/java/org/elasticsearch/index/search/NestedHelper.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/search/NestedHelper.java:65]: agreed, either assert false or throw unchecked IO

**Review by javanna — APPROVED** [review]
LGTM besides that small comment about exception handling

File: server/src/main/java/org/elasticsearch/index/search/NestedHelper.java (hunk 1 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/search/NestedHelper.java:65]: We cannot get an IOException, its never thrown and the Exception will be removed in Lucene, see https://github.com/apache/lucene/pull/13701
As for the AssertionException: this is quite common elsewhere when e.g. iterating BytesRefIterator [here in MessageDigest](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/hash/MessageDigests.java#L144) or [here in Netty4Utils](https://github.com/elastic/elasticsearch/blob/main/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Utils.java#L114).

That said, I don't care, will change it to unchecked IO fwiw

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/search/NestedHelper.java]: > As for the AssertionException: this is quite common elsewhere when e.g. iterating BytesRefIterator

I disagree with those usages as well. Anything that has a "This should never happen" in the line makes me really uncomfortable. Its like asking for things to break :D. 

If we want to have assertions, we should use them. Then either ignore the thrown exception or throw a real exception with a real description encapsulating the actually thrown exception.

**Review by benwtrent — APPROVED** [review]

**Comment by cbuescher** [comment]
CI errors were unrelated and also reproducible [on main](https://github.com/elastic/elasticsearch/issues/113185).

elasticmachine run elasticsearch-ci/part-4

**Comment by cbuescher** [comment]
elasticmachine run elasticsearch-ci/part-4

**Pull request merged**